### PR TITLE
Improve spectrum analysis

### PR DESCRIPTION
**Pull request by joleroi**
This PR introduces a new background method to the spectral fitting utilities: OFF vectors from Background models.

The appropriate statistics in Sherpa (WStat vs. cash) is not yet available. Default stat is set to CASH

* [x] add method to compute OFF vec from background model
* [x] add debug plots
* [x] add functionality to exclude runs without sufficient number of reflected regions
* [x] add copy function to SpectrumAnalysis (to try different bkg methods)

### CODE REVIEW

**Comment by joleroi** [comment]
I would like to merge this today. Comments welcome

**Comment by joleroi** [comment]
@cdeil : one question - in gammapy-extra/datasets/hess-crab4_pha I have some example PHA data. In order to use this for tests I set the reference to the RMF and ARF files in the PHA header relative to $GAMMAPY_EXTRA.
Now I get this error
https://travis-ci.org/gammapy/gammapy/jobs/97637674#L1431

Can I in principle use gammapy-extra this way or does it only work via gammapy.datasets.gammapy-extra. If the latter is the case, I don't know how to fix this test

**Comment by joleroi** [comment]
@cdeil: I found the error, it was a bug on my side

**Comment by cdeil** [comment]
@joleroi - Rebase on master, please.

**Comment by joleroi** [comment]
Test still fail because of https://github.com/gammapy/gammapy/issues/408 . I guess that used to happen only locally, but I guess now it's also a problem for travis (which is a good thing I guess). I have no idea how to attack this, though.

**Comment by joleroi** [comment]
Any comments before I merge?

File: gammapy/region/circle.py [hunk 1]
```diff
@@ -132,6 +132,21 @@ def __init__(self, pos, radius):
         self.pos = SkyCoord(pos)
         self.radius = Angle(radius)
 
+    @property
+    def area(self):
+        """Circle Area
+        """
```
Comment (cdeil) [inline gammapy/region/circle.py]: This formula isn't obvious.

Please leave a comment that it's the spherical cap area and a link to:
https://en.wikipedia.org/wiki/Solid_angle#Cone.2C_spherical_cap.2C_hemisphere
https://en.wikipedia.org/wiki/Versine#Definitions
Maybe even putting this in the docstring is better than as a code comment?

This should be equivalent and slightly simpler / faster, no?
```python
val = 2 * np.pi * (1 - np.cos(self.radius))
```

File: gammapy/region/tests/test_circle.py [hunk 2]
```diff
@@ -50,3 +52,11 @@ def test_sky_to_pix_to_sky(self):
         assert_allclose(sky.pos.l, sky2.pos.l)
         assert_allclose(sky.pos.b, sky2.pos.b)
         assert_allclose(sky.radius, sky2.radius)
+
+    def test_area(self):
+        pos = SkyCoord.from_name('crab')
```
Comment (cdeil) [inline gammapy/region/tests/test_circle.py]: The `SkyCoord.from_name` method makes a query to the internet SESAME service.
This is better avoided in tests.

You could hard-code the Crab position or choose some random other position, no?
```python
>>> from astropy.coordinates import SkyCoord
>>> SkyCoord.from_name('crab')
<SkyCoord (ICRS): (ra, dec) in deg
    (83.633083, 22.0145)>
```

File: gammapy/region/tests/test_circle.py [hunk 3]
```diff
@@ -50,3 +52,11 @@ def test_sky_to_pix_to_sky(self):
...
+        rad = Angle('1.3451 deg')
+        reg = SkyCircleRegion(pos, rad)
+        desired = (rad ** 2 * np.pi).to('steradian')
```
Comment (cdeil) [inline gammapy/region/tests/test_circle.py]: Leave a comment that this is the small angle approximation, and the method is the spherical cap area.

File: gammapy/spectrum/spectrum_analysis.py [hunk 4]
```diff
@@ -1,12 +1,18 @@
...
 import logging
+from astropy import wcs
+
 import numpy as np
 from astropy.coordinates import Angle, SkyCoord
 from astropy.extern import six
-from ..image import ExclusionMask
+from astropy.wcs.utils import skycoord_to_pixel
+
+from gammapy.extern import pathlib
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: Let's use relative explicit imports everywhere. I.e. here:
```python
from ..extern.pathlib import Path
```
I think this import is superfluous, no?
```python
from gammapy.extern import pathlib
```

File: gammapy/spectrum/spectrum_analysis.py [hunk 5]
```diff
@@ -40,6 +46,33 @@ def main(args=None):
     run_spectral_fit_using_config(config)
 
 
+def filter_reflected_regions_analysis(analysis, n_min):
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: Make this a method on `gammapy.spectrum.SpectrumAnalysis`?

I know having more modularity and not aggregating all functionality in `*Analysis` is good,
but if the standalone function takes one single `*Analysis` object as input to do something,
there's no advantage over making it a method, no?
(I'd claim the disadvantage is that it's harder to find / understand for users if such things are split into separate functions in the `gammapy.spectrum` namespace.)

File: gammapy/spectrum/spectrum_analysis.py [hunk 6]
```diff
@@ -40,6 +46,33 @@ def main(args=None):
...
+    Parameters
+    ----------
+    analysis : `~gammapy.spectrum.SpectrumAnalysis`
+        Spectrum analysis with reflected regions background method
+    n_min : int
+        Minimum number of reflected regions
+
+    Returns
+    -------
+    mask : `np.array`
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: I usually use `mask` for boolean mask arrays and `idx` or `index` for index arrays.
I think that's also the case in numpy / scipy / ....

So may I suggest you call this `idx` and call [numpy.nonzero](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.nonzero.html) here instead of [numpy.where](http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.where.html)?

File: gammapy/spectrum/spectrum_analysis.py [hunk 7]
```diff
@@ -85,37 +119,106 @@ def __init__(self, datastore, obs, on_region, exclusion, bkg_method=None,
...
+    def alpha(self):
+        """`np.array` of all exposure ratios between ON and OFF regions
+        """
+        return np.array([obs.alpha for obs in self.observations])
+
+    @property
+    def total_alpha(self):
+        """Averaged exposure ratio between ON and OFF regions
+
+        Alpha for all observations is calculated as :
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: Did you want to add a formula here?

File: gammapy/spectrum/spectrum_analysis.py [hunk 8]
```diff
@@ -385,44 +534,71 @@ def write_ogip(self, phafile=None, bkgfile=None, rmffile=None, arffile=None,
...
+
+        self.on_vector.write(str(outdir/phafile), bkg=str(bkgfile), arf=str(arffile),
+                             rmf=str(rmffile), clobber=clobber)
+        self.off_vector.write(str(outdir/bkgfile), clobber=clobber)
+        self.effective_area.write(str(outdir/arffile), energy_unit='keV',
+                                  effarea_unit='cm2', clobber=clobber)
+        self.energy_dispersion.write(str(outdir/rmffile), energy_unit='keV',
+                                     clobber=clobber)
+
+    def plot_exclusion_mask(self, size=None, **kwargs):
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: Pass `ax` so that it's possible to easily create this plot as one of several panels in a larger Figure?

File: gammapy/spectrum/spectrum_analysis.py [hunk 9]
```diff
@@ -434,18 +610,30 @@ class SpectralFit(object):
...
-    def __init__(self, pha, bkg=None, arf=None, rmf=None):
+    DEFAULT_STAT = 'cash'
+
+    def __init__(self, pha, bkg=None, arf=None, rmf=None, stat=DEFAULT_STAT):
 
         self.pha = [make_path(f) for f in pha]
         self._model = None
         self._thres_lo = None
         self._thres_hi = None
+        self.statistic = stat
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: Call it consistently `statistic` or `stat`?
I think Sherpa uses `stat` in most places, no?

File: gammapy/spectrum/spectrum_analysis.py [hunk 10]
```diff
@@ -480,11 +668,38 @@ def model(self, model, name=None):
                 raise ValueError("Undefined model string: {}".format(model))
 
         if not isinstance(model, sherpa.models.ArithmeticModel):
-            raise ValueError("Only sherpa models are supported at the moment")
+            raise ValueError("Only sherpa models are supported")
 
         self._model = model
 
     @property
+    def statistic(self):
```
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: In `__init__` you called the data member `statistic`.
Now you're creating a property `statistic` that accesses a data member `_statistic`.
Is this a bug?

File: gammapy/spectrum/tests/test_counts_spectrum.py [hunk 11]
```diff
@@ -1,10 +1,17 @@
 # Licensed under a 3-clause BSD style license - see LICENSE.rst
-from __future__ import absolute_import, division, print_function, unicode_literals
+from __future__ import absolute_import, division, print_function, \
+    unicode_literals
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_counts_spectrum.py]: Let's keep this on one line.
We have > 100 of those lines in Gammapy, and I put them all on one line.

File: gammapy/spectrum/tests/test_counts_spectrum.py [hunk 12]
```diff
@@ -1,10 +1,17 @@
...
+
+import numpy as np
 from numpy.testing import assert_equal
+
+from gammapy.datasets import gammapy_extra
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_counts_spectrum.py]: Use relative explicit imports consistently within Gammapy.

File: gammapy/spectrum/tests/test_spectrum_analysis.py [hunk 13]
```diff
@@ -60,28 +62,28 @@ def test_spectral_fit(tmpdir):
     fit.energy_threshold_low = '100 GeV'
     fit.energy_threshold_high = '10 TeV'
     fit.run(method='sherpa')
-    assert_allclose(fit.model.gamma.val, 2.0, rtol=1e-1)
+    assert_allclose(fit.model.gamma.val, 2.23, rtol=1e-1)
```
Comment (cdeil) [inline gammapy/spectrum/tests/test_spectrum_analysis.py]: It's great to now have first high-level tests passing on travis-ci:
https://travis-ci.org/gammapy/gammapy/jobs/101052536#L1449

For spectral index, I'd find `atol` more intuitive than `rtol`. Can you please change to `atol` here?

**Comment by cdeil** [comment]
I've left a few inline comments.

Maybe change the PR title from "OFF vector from cube background model" to "Improve spectrum analysis" since this PR now does several important improvements?

Also this comment in the PR description is cryptic:
> The appropriate statistics in Sherpa (WStat vs. cash) is not yet available. Default stat is set to CASH

When using a background spectral template extracted from stacked off runs, using CASH is the right thing to do, and it's available, and you're using it. Maybe rephrase?

If you have time, please look over the coverage of the files you've worked on:
https://coveralls.io/builds/4663382
Maybe this could be improved a bit?
https://coveralls.io/builds/4663382/source?filename=gammapy%2Fscripts%2Fspectrum_pipe.py

Wouldn't it be better to split the `SpectrumAnalysis.make_comparison_plot` functionality out into a standalone function or class? I.e. have `SpectrumAnalysis` only be responsible for running one analysis, have a way to serialise the results, and then have separate functions / classes to compare analyses?
Of course it's OK to do this in future PRs ... I just saw that this here for the first time:
https://coveralls.io/builds/4663382/source?filename=gammapy%2Fscripts%2Fspectrum_pipe.py#L139

**Comment by joleroi** [comment]
Thank for the comments, I will go trough them today.  About your general comments:
* Concerning the statistics issue: At the moment ONLY cash is availabe for all spectral fits. The Version of sherpa I have installed (4.7) does not have WStat, and I don't manage to install the head version
```
checking for flex... flex
checking lex output file root... configure: error: cannot find output from flex; giving up
configure: error: ./configure failed for region-4.7
```
This looks related https://gist.github.com/cdeil/39e9404e91954cb67270. Suggestions?

*  Continued ...

> When using a background spectral template extracted from stacked off runs, using CASH is the right thing to do

Yes and I don't know, the background model gives non-integer couns, sherpa only understands PHA (i.e. integer counts). I am happy for any suggestions what to do :grin: 

* Make comparison plot

> Wouldn't it be better to split the SpectrumAnalysis.make_comparison_plot functionality out into a standalone function or class?

Of course this is just something I hacked together. It will go away soon!

**Comment by cdeil** [comment]
@joleroi - Could you please make EDISP2D spec a priority because of the upcoming telcon (https://github.com/gammapy/gamma-astro-data-formats/issues/16#issuecomment-170451773)?

Concerning the Sherpa build issue and how to set the non-integer expected counts background model in Sherpa ... I don't know, but I could come by later today and we try to figure it out (or maybe also Peter or Regis can help with the spectrum background model issue).

File: gammapy/spectrum/spectrum_analysis.py (hunk 4 above)
Comment (joleroi) [inline gammapy/spectrum/spectrum_analysis.py]: @cdeil : Do you know how to teach PyCharm to make import relative when optimizing imports?

Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: > @cdeil : Do you know how to teach PyCharm to make import relative when optimizing imports?

I don't think that's possible. It probably would have to be added as an option under Editor -> General -> Auto Import -> Python -> Preferred import style.
The only feature request I could find is this, which at a very quick glance looks different:
https://youtrack.jetbrains.com/issue/PY-9987

I'd also like to have that option ... can you take the time to file a feature request in the PyCharm tracker? Or should I?

File: gammapy/spectrum/spectrum_analysis.py (hunk 8 above)
Comment (joleroi) [inline gammapy/spectrum/spectrum_analysis.py]: I had a discussion about this with Axel, the problem is that I need a WCS axis object, right?
This object has to know the transformation defined in the Exclusion mask header apparently due to some binning issue. So, how can I pass an axis and then give it the WCS info from the exclusion mask?

File: gammapy/spectrum/spectrum_analysis.py (hunk 10 above)
Comment (joleroi) [inline gammapy/spectrum/spectrum_analysis.py]: Yes

File: gammapy/spectrum/spectrum_analysis.py (hunk 8 above)
Comment (cdeil) [inline gammapy/spectrum/spectrum_analysis.py]: I vaguely remember discussing this in the past as well and don't know if we found a good solution.
Maybe split this out in a separate issue and leave as-is for now?

**Comment by joleroi** [comment]
@cdeil Do you know the the tests behave so strangely?

**Comment by cdeil** [comment]
Do you mean on the CI servers?

You have to rebase this branch:

screen shot 2016-01-14 at 09 46 00

If you mean local testing, I could come by later today and help debug it.

**Comment by joleroi** [comment]
I meant on the CI servers. Was it always the case that the need for rebase made the test not run? But anyhow, thanks!

**Comment by joleroi** [comment]
Can I merge?

**Comment by cdeil** [comment]
:+1:  to merge!

(I didn't review this in detail or try it out because I don't have time at the moment. So I might have comments in the future.)
